system/include/SSE/xmmintrin.h

```diff
@@ -0,0 +1,709 @@
+{
+  return __builtin_shufflevector(__a, wasm_f32x4_make(((float*)__p)[0], ((float*)__p)[1], 0.f, 0.f), 0, 1, 4, 5);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_loadr_ps(const float *__p)
+{
+  __m128 __v = _mm_load_ps(__p);
+  return __builtin_shufflevector(__v, __v, 3, 2, 1, 0);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_loadu_ps(const float *__p)
+{
+  return wasm_v128_load(__p);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_load_ps1(const float *__p)
+{
+  return wasm_v32x4_load_splat(__p);
+}
+#define _mm_load1_ps _mm_load_ps1
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_load_ss(const float *__p)
+{
+  return wasm_f32x4_make(*__p, 0.f, 0.f, 0.f);
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_storel_pi(__m64 *__p, __m128 __a)
+{
+  *__p = (__m64)__a;
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_storeh_pi(__m64 *__p, __m128 __a)
+{
+  *__p = (__m64)__builtin_shufflevector(__a, __a, 6, 7, 2, 3); // _mm_movehl_ps pattern
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_store_ps(float *__p, __m128 __a)
+{
+  *(__m128 *)__p = __a;
+}
+// No NTA cache hint available.
+#define _mm_stream_ps _mm_store_ps
+
+#define _MM_HINT_T0 3
+#define _MM_HINT_T1 2
+#define _MM_HINT_T2 1
+#define _MM_HINT_NTA 0
+// No prefetch available, dummy it out.
+static __inline__ void __attribute__((__always_inline__))
+_mm_prefetch(const void *__p, int __i)
+{
+  ((void)__p);
+  ((void)__i);
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_sfence(void)
+{
+  __sync_synchronize(); // Emscripten/SharedArrayBuffer has only a full barrier instruction, which gives a stronger guarantee.
+}
+
+#define _MM_SHUFFLE(w, z, y, x) (((w) << 6) | ((z) << 4) | ((y) << 2) | (x))
+
+// This is defined as a macro because __builtin_shufflevector requires its
+// mask argument to be a compile-time constant.
+#define _mm_shuffle_ps(a, b, mask) \
+  ((__m128)__builtin_shufflevector((__f32x4)(a), (__f32x4)(b), \
+                                  (((mask) >> 0) & 0x3) + 0, \
+                                  (((mask) >> 2) & 0x3) + 0, \
+                                  (((mask) >> 4) & 0x3) + 4, \
+                                  (((mask) >> 6) & 0x3) + 4))
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_storer_ps(float *__p, __m128 __a)
+{
+  _mm_store_ps(__p, _mm_shuffle_ps(__a, __a, _MM_SHUFFLE(0, 1, 2, 3)));
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_store_ps1(float *__p, __m128 __a)
+{
+  _mm_store_ps(__p, _mm_shuffle_ps(__a, __a, _MM_SHUFFLE(0, 0, 0, 0)));
+}
+#define _mm_store1_ps _mm_store_ps1
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_store_ss(float *__p, __m128 __a)
+{
+    struct __unaligned {
+      float __v;
+    } __attribute__((__packed__, __may_alias__));
+    ((struct __unaligned *)__p)->__v = ((__f32x4)__a)[0];
```

Comment:
`__f32x4` and friends are not meant to be exported from wasm_simd128.h, but I suppose if you copied the definitions to this file you would get errors about multiple typedefs of the same thing. Maybe they should be macros in wasm_simd128.h so they can be undefined at the end? Anyway, I think this is fine for now but it might be something we want to change in the future.

Reply:
Yeah, I hope it'll be fine - if upstream changes, we should be able to adapt quite easily.